src/cls/rbd/cls_rbd_types.h

```diff
@@ -216,6 +217,140 @@ struct GroupSpec {
 
 WRITE_CLASS_ENCODER(GroupSpec);
 
+enum SnapshotNamespaceType {
+  SNAPSHOT_NAMESPACE_TYPE_USER = 0,
+  SNAPSHOT_NAMESPACE_TYPE_GROUP = 1
+};
+
+struct UserSnapshotNamespace {
+  static const uint32_t SNAPSHOT_NAMESPACE_TYPE = SNAPSHOT_NAMESPACE_TYPE_USER;
+
+  UserSnapshotNamespace() {}
+
+  void encode(bufferlist& bl) const {}
+  void decode(bufferlist::iterator& it) {}
+
```

Comment:
I tried to implement it using inheritance and couldn't reconcile all types of constructors that are provided by boost::variant. It looks like the solution is to have a wrapper type like NotifyMessage wraps Payload and EventEntry wraps Event type. Those wrappers implement all the required methods.